### PR TITLE
Iosapp

### PR DESCRIPTION
**Pull request by robbiecarlton**

### CODE REVIEW

**Comment by kkireyev** [comment]
Robbie from Authentic World? You work for Hylo now? Awesome!

File: api/models/Comment.js [hunk 1]
```diff
@@ -93,5 +96,36 @@ module.exports = bookshelf.Model.extend({
     })
 
   },
+  
+  sendPushNotification: function(recipientId, comment) {
+
+    return Promise.join(
+      User.find(recipientId),
+      comment.load(['user', 'post', 'post.communities', 'post.creator'])
```
Comment (levity) [inline api/models/Comment.js]: needs to be passed {transacting: trx}

File: api/controllers/PushNotificationController.js [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+	token: req.param("deviceToken"),
+	user_id: req.session.userId
+      })
+        .fetch()
+	.then(function (device) {
+          device.save({
+            badge_no: req.param("badgeNo") || 0
+          })
+        })
+        .then(function() {
```
Comment (levity) [inline api/controllers/PushNotificationController.js]: this could be `then(() => res.ok({result: 'updated'}))`

File: api/models/Comment.js [hunk 3]
```diff
@@ -93,5 +96,36 @@ module.exports = bookshelf.Model.extend({
...
+
+        if (!comment) return;
+
+        var seed = comment.relations.post,
+            commenter = comment.relations.user,
+            creator = seed.relations.creator;
+
+        if (seed.relations.communities) {
+          var community = seed.relations.communities.models[0],
+              text = RichText.qualifyLinks(comment.get('comment_text')),
```
Comment (levity) [inline api/models/Comment.js]: not being used

File: api/models/Device.js [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+  tableName: 'devices',
+    
+  user: function () {
+    return this.belongsTo(User, "user_id");
+  },
+
+  sendPushNotification: function(alert, url) {
+    var badge_no = this.get('badge_no') + 1;
+    this.set("badge_no", badge_no);
+    this.save();
```
Comment (levity) [inline api/models/Device.js]: save() is async so it needs a then

File: api/models/Qdpush.js [hunk 5]
```diff
@@ -0,0 +1,7 @@
+var format = require('util').format;
+
+module.exports = bookshelf.Model.extend({
+  tableName: 'queued_pushes',
```
Comment (levity) [inline api/models/Qdpush.js]: rename to push_notifications

File: api/models/User.js [hunk 6]
```diff
@@ -215,6 +231,12 @@ module.exports = bookshelf.Model.extend({
       unsalted = decrypted.replace(new RegExp('^' + process.env.MAILGUN_EMAIL_SALT), '');
 
     return unsalted;
+  },
+  
+  sendPushNotification: function(userId, alert, url) {
+    return User.find(userId)
+      .fetch()
+      .sendPushNotification(alert, url)
```
Comment (levity) [inline api/models/User.js]: semicolon

File: sendqdpushes.js [hunk 7]
```diff
@@ -0,0 +1,64 @@
+var skiff = require('./lib/skiff'); // this must be first
+var apn = require('apn'),
+    sails = skiff.sails;
+
+var apnOptions = {passphrase: "kniujhYY&u",
```
Comment (levity) [inline sendqdpushes.js]: define in `.env` and refer to it as `process.env.MY_VAR_NAME`

File: sendqdpushes.js [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+          note.alert = qdpush.get("alert");
+          note.payload = JSON.parse(qdpush.get("payload"));
+          
+          apnConnection.pushNotification(note, myDevice);
+          
+          console.log(qdpush.get("payload"));
+
+          qdpush.set("time_sent", (new Date()).toISOString());
+
+          qdpush.save();
```
Comment (levity) [inline sendqdpushes.js]: async

File: sendqdpushes.js [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+
+skiff.lift({
+  start: function(argv) {
+
+    Qdpush.query(function(qb) {
+      qb.whereNull("time_sent")
+    })
+      .fetchAll()
+      .then(function (qdpushes) {
+        return qdpushes.map(function (qdpush) {
```
Comment (levity) [inline sendqdpushes.js]: Promise.map

**Comment by levity** [comment]
Hi @kkireyev :)

**Comment by kkireyev** [comment]
Hi, my Hylo homies!

**Comment by robbiecarlton** [comment]
Failing a test, any insight into that Lawrence?
Otherwise, made the changes you asked for.

**Comment by edwardwest** [comment]
Hey @kkireyev ! glad to see you here!  :)

We were sprinting on Projects (now live, check it out! :)  and we are going
to be incorporating your work on the Serendipity Engine ***very soon.**  (
@levity has had a lot on his plate.)

We are super stoked to have @robbiecarlton join the team-- he is an awesome
guy, and it was fun to learn that he's also an epic developer.  :)

@kkireyev -- do you have time for lunch next week at Hub?

On Thursday, May 28, 2015, Lawrence Wang <notifications@github.com
<javascript:_e(%7B%7D,'cvml','notifications@github.com');>> wrote:

> Hi @kkireyev <https://github.com/kkireyev> :)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Hylozoic/hylo-node/pull/15#issuecomment-106570347>.
>


-- 
_______________________
*Edward West*
mobile 415.225.7335
Facebook <http://www.facebook.com/edwardjwest>  Twitter
<http://www.twitter.com/edwardwest>  LinkedIn
<http://www.linkedin.com/in/edwardwest>

**Comment by levity** [comment]
Don't worry about the tests; the testing environment has some environment
variables that need to be set, but CircleCI disables them when running
tests for forked repos because they could contain sensitive info.

Can you do another once-over for style (semicolons & indentation) and
remove any debugging print statements?

**Comment by levity** [comment]
But if you wanted to go in the other direction and omit all unnecessary semicolons, according to http://blog.izs.me/post/2353458699/an-open-letter-to-javascript-leaders-regarding, I could be game

**Comment by robbiecarlton** [comment]
@levity Cleaned up the semi-colons and removed debugging print statements

@kkireyev Wasup! :smile:

**Comment by levity** [comment]
work continues at https://github.com/Hylozoic/hylo-node/pull/17

**Pull request closed without merging**